agent/php_nrini.c

```diff
@@ -3047,8 +3046,7 @@ static int nr_ini_displayer_global(zend_ini_entry* ini_entry,
    * If there is no value, then don't print anything for the "special" ini
    * settings.
    */
-  if ((NULL == PHP_INI_ENTRY_VALUE(ini_entry))
-      || (0 == PHP_INI_ENTRY_VALUE_LEN(ini_entry))) {
+  if (0 == PHP_INI_ENTRY_VALUE_LEN(ini_entry)) {
```

Comment:
These are both errors associated with the warning that throw the 
`error: the comparison will always evaluate as 'false'`
These are the definitions for those macros:
```c
#define PHP_INI_ENTRY_VALUE(ie) (ie)->value->val
#define PHP_INI_ENTRY_VALUE_LEN(ie) (ie)->value->len
```
So no apparent `NULL` checking. However, if you expand the context of that function, it becomes clear that `ini_entry` cannot be `NULL` at that point in the code, hence why the compiler complains (I would guess).
```c
static int nr_ini_displayer_global(zend_ini_entry* ini_entry,
                                   int* module_number,
                                   zend_hash_key* key NRUNUSED TSRMLS_DC) {
  if (ini_entry->module_number != *module_number) {
    return ZEND_HASH_APPLY_KEEP;
  }

  if (ini_entry->modifiable & PHP_INI_PERDIR) {
    return ZEND_HASH_APPLY_KEEP;
  }

  /*
   * If there is no value, then don't print anything for the "special" ini
   * settings.
   */
```
By this point, the program would have errored  out by attempting to reference illegal memory locations `module_number` & `modifiable` if `ini_entry` was `NULL`

Reply:
Ok. So this is because of how `(ie)->value->val` is defined: `char              val[1]`, which means `val` is never NULL.

Reply:
Do we know that `val` is defined that way for all PHP versions we compile against?
If in the past or future it is changed to just a pointer definition we will be missing lots of NULL checks.

**I think there are several examples in this PR of removing code which could be better addressed by the pragma solution but I haven't investigated them all.**

I think it could be better to wrap these checks with:
```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored <WARNING>
  foo(b);                       /* no diagnostic for this one */
#pragma GCC diagnostic pop

```
where `<WARNING>` is whatever "-W" warning that caused the trouble.

We have examples in the agent already like this:
https://github.com/newrelic/newrelic-php-agent/blob/cebc9fa0097393cb75ecf2d292ddb26ec0a13e02/agent/php_agent.c#L947

See https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html for more details.

Reply:
Looks like `val` is always defined that way for zend_string
Here is php 8.1 https://github.com/php/php-src/blob/PHP-8.1/Zend/zend_types.h#L350

for php 5.5
zend_string does not exist.

for php 8.1, 
but `value` is a pointer to zend_string and can theroetically be checked for NULL.
https://github.com/php/php-src/blob/master/Zend/zend_ini.h#L53
also for php 5.5, value (https://github.com/php/php-src/blob/PHP-5.5/Zend/zend_ini.h#L73) is simply a pointer to char and can be checked or null.

So in this case, we see why it's handled differently:
```
#ifdef PHP7
#define PHP_INI_ENTRY_NAME(ie) (ie)->name->val
#define PHP_INI_ENTRY_NAME_LEN(ie) (ie)->name->len + 1
#define PHP_INI_ENTRY_ORIG_VALUE(ie) (ie)->orig_value->val
#define PHP_INI_ENTRY_ORIG_VALUE_LEN(ie) (ie)->orig_value->len
#define PHP_INI_ENTRY_VALUE(ie) (ie)->value->val
#define PHP_INI_ENTRY_VALUE_LEN(ie) (ie)->value->len
#else
#define PHP_INI_ENTRY_NAME(ie) (ie)->name
#define PHP_INI_ENTRY_NAME_LEN(ie) (ie)->name_length
#define PHP_INI_ENTRY_ORIG_VALUE(ie) (ie)->orig_value
#define PHP_INI_ENTRY_ORIG_VALUE_LEN(ie) (ie)->orig_value_length
#define PHP_INI_ENTRY_VALUE(ie) (ie)->value
#define PHP_INI_ENTRY_VALUE_LEN(ie) (ie)->value_length
#endif
```


Did you try running this compiler with PHP 5.5?

Probably best to put the logic change  seperated similar to the PHP_INI_ENTRY* logic, where one way is the new way here, and the other way still follows the old logic for PHP 5.5 and 5.6.

Reply:
I've addressed this using the `#pragma` directive as suggested by @mfulb